onnxruntime/core/providers/xnnpack/math/gemm.cc

```diff
@@ -0,0 +1,204 @@
+// Copyright(c) Microsoft Corporation.All rights reserved.
+// Licensed under the MIT License.
+
+#include "gemm.h"
+#include "core/framework/transpose_helper.h"
+#include "core/providers/utils.h"
+
+namespace onnxruntime {
+namespace xnnpack {
+
+// Todo -
+// 1. Integrate activation layers - Cliping & Relu 
+// 2. Enable C matrix broadcasting - reuse "GemmBroadcastBias" function / logic
+// 3. Enable Quant ops
+// 4. Review possible consolidation of MatMul & Gemm 
+//
+
+bool Gemm::IsGemmOnnxNodeSupported(const NodeUnit& node_unit, const GraphViewer& graph) {
+  bool supported = false;
+  const onnxruntime::Node& node = node_unit.GetNode();
+
+  // use do {} while(false) so it's easier to set a breakpoint on the return
+  do {
+    ConstPointerContainer<std::vector<NodeArg*>> input_defs = node.InputDefs();
+
+    const auto alpha = node.GetAttributes().find("alpha");
+    if ((*alpha).second.f() != 1.0) break;
+
+    const auto beta = node.GetAttributes().find("beta");
+    if ((*beta).second.f() != 1.0) break;
+
+
+    const NodeArg* A_arg = input_defs[0];
+    const NodeArg* B_arg = input_defs[1];
+    const NodeArg* C_arg = input_defs.size() == 2 ? nullptr : input_defs[2];
+
+    // we only support float currently
+    const auto* A_type = A_arg->TypeAsProto();
+
+    if (A_type == nullptr || 
+        A_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+      break;
+    }
+
+    // B & C matrices must be constant
+    if (!graph.IsConstantInitializer(B_arg->Name(), true)) {
+      break;
+    }
+
+    if (input_defs.size() == 3 && !graph.IsConstantInitializer(C_arg->Name(), true)) {
+      break;
+    }
+
+    // making sure we are dealing with MatMul
+    const ONNX_NAMESPACE::TensorShapeProto* A_shape = A_arg->Shape();
+    const ONNX_NAMESPACE::TensorShapeProto* B_shape = B_arg->Shape();
+    const ONNX_NAMESPACE::TensorShapeProto* C_shape = C_arg->Shape();
+
+    if (!A_shape || A_shape->dim_size() >= 3) {
+      break;
+    }
+
+    if (!B_shape || B_shape->dim_size() >= 3) {
+      break;
+    }
+
+    if (!C_shape || C_shape->dim_size() >= 3) {
+      break;
+    }
+    
+    if (input_defs.size() == 3 && (C_shape->dim(0).dim_value() != B_shape->dim(1).dim_value() && C_shape->dim(0).dim_value() != B_shape->dim(0).dim_value())) {
+      break;
+    }
+
+    supported = true;
+
+  } while (false);
+
+  return supported;
+}
+
+Gemm::Gemm(const OpKernelInfo& info) : GemmBase(info), XnnpackKernel(info) {
+  const auto& node{Node()};
+
+  info.GetAttrOrDefault<float>("alpha", &alpha_, 1.f);
+  info.GetAttrOrDefault<float>("beta", &beta_, 1.f);
+
+  const auto& input_defs = node.InputDefs();
+  const auto* shapeA = input_defs[0]->Shape();
+  const auto* shapeB = input_defs[1]->Shape();
+
+  numberOfInputs_ = input_defs.size();
+
```

Comment:
Maybe this is the best place to check if a) the optional 3rd input is present, and if so Exists() is true or not. All other code could use this value when deciding if there's a bias input to handle